src/network/networkd-sysctl.c

```
@@ -5,15 +5,196 @@
+}
+
+void sysctl_add_monitor(Manager *manager) {
+        _cleanup_(sysctl_monitor_bpf_freep) struct sysctl_monitor_bpf *obj = NULL;
+        _cleanup_(bpf_link_freep) struct bpf_link *sysctl_link = NULL;
+        _cleanup_(rb_freep) struct ring_buffer *sysctl_buffer = NULL;
+        _cleanup_close_ int cgroup_fd = -EBADF;
+        int r;
+
+        assert(manager);
+
+        r = dlopen_bpf();
+        if (r < 0) {
+                log_info("sysctl monitor disabled, as BPF support is not available.");
+                return;
+        }
+
+        cgroup_fd = open_rootcg();
+        if (cgroup_fd < 0)
+                return;
+
+        obj = sysctl_monitor_bpf__open_and_load();
+        if (obj == NULL) {
+                log_warning_errno(errno, "Failed to load sysctl monitor BPF program: %m");
+                return;
+        }
+
+        sysctl_link = bpf_program__attach_cgroup(obj->progs.sysctl_monitor, cgroup_fd);
+        r = bpf_get_error_translated(sysctl_link);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to attach sysctl monitor BPF program to cgroup: %m");
+                return;
+        }
+
+        sysctl_buffer = ring_buffer__new(
+                        bpf_map__fd(obj->maps.written_sysctls),
+                        sysct_write_event, &manager->sysctl_shadow, NULL);
+        if (!sysctl_buffer) {
+                log_warning_errno(errno, "Failed to create ring buffer: %m");
+                return;
+        }
+
+        r = sd_event_add_io(manager->event, &manager->sysctl_event_source,
+                        ring_buffer__epoll_fd(sysctl_buffer), EPOLLIN, on_ringbuf_io, sysctl_buffer);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to watch sysctl event ringbuffer: %m");
+                return;
+        }
+
+        manager->sysctl_link = TAKE_PTR(sysctl_link);
+        manager->sysctl_skel = TAKE_PTR(obj);
+        manager->sysctl_buffer = TAKE_PTR(sysctl_buffer);
+
+        return;
```

this line is entirely redundant.

that said, i still would suggest that sysctl_add_monitor() and sysctl_remove_monitor() should return a numeric error code (i.e. `int` rather than `void`).

We typically prefer that functions return errors, and if those errors shall be ignored (which often is fine) that it's the caller that does it, not already the callee.

---

rearranging things like that means you can use use in the eove:

```c
if (r < 0)
        return log_warning_errno(r, …);
```

instead of:

```
if (r < 0) {
        log_warning_errno(r, …);
        return;
}
```

i.e. error handling takes only 2 lines instead of of 4 per function call